src/Strategy/ManifestStrategy.php

```diff
@@ -0,0 +1,178 @@
+        if (!$this->allowMajor) {
+            $versions = $this->filterByLocalMajorVersion($versions);
+        }
+
+        $versionParser = new VersionParser($versions);
+
+        $mostRecent = $versionParser->getMostRecentStable();
+
+        // Look for unstable updates if explicitly allowed, or if the local
+        // version is already unstable and there is no new stable version.
+        if ($this->allowUnstable || ($versionParser->isUnstable($this->localVersion) && version_compare($mostRecent, $this->localVersion, '<'))) {
+            $mostRecent = $versionParser->getMostRecentAll();
+        }
+
+        return version_compare($mostRecent, $this->localVersion, '>') ? $mostRecent : false;
+    }
+
+    /**
+     * Get available versions to update to.
+     *
+     * @return array
+     *   An array keyed by the version name, whose elements are arrays
+     *   containing version information ('name', 'sha1', and 'url').
+     */
+    private function getAvailableVersions()
+    {
+        if (!isset($this->availableVersions)) {
+            $this->availableVersions = array();
+            foreach ($this->getManifest() as $key => $item) {
+                if ($missing = array_diff(self::$requiredKeys, array_keys($item))) {
+                    throw new \RuntimeException(sprintf('Manifest item %s missing required key(s): %s', $key, implode(',', $missing)));
+                }
+                $this->availableVersions[$item['version']] = $item;
+            }
+        }
+
+        return $this->availableVersions;
+    }
+
+    /**
+     * Download and decode the JSON manifest file.
+     *
+     * @return array
+     */
+    private function getManifest()
+    {
+        if (!isset($this->manifest)) {
+            $manifestContents = file_get_contents($this->manifestUrl);
+            if ($manifestContents === false) {
+                throw new \RuntimeException(sprintf('Failed to download manifest: %s', $this->manifestUrl));
+            }
+
+            $this->manifest = (array) json_decode($manifestContents, true);
+            if (json_last_error() !== JSON_ERROR_NONE) {
+                throw new JsonParsingException(
+                    'Error parsing manifest file'
+                    . (function_exists('json_last_error_msg') ? ': ' . json_last_error_msg() : '')
+                );
+            }
+        }
+
+        return $this->manifest;
+    }
+
+    /**
+     * Filter a list of versions to those that match the current local version.
+     *
+     * @param string[] $versions
+     *
+     * @return string[]
+     */
+    private function filterByLocalMajorVersion(array $versions)
+    {
+        list($localMajorVersion, ) = explode('.', $this->localVersion, 2);
+        return array_filter($versions, function ($version) use ($localMajorVersion) {
+            list($majorVersion, ) = explode('.', $version, 2);
+            return $majorVersion === $localMajorVersion;
```

Comment:
I personally like (but maybe @humbug/core team has another opinion on that) to have a blank line before a return/exit statement. It makes it easier to identify